pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,473 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package photon
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/vmware/photon-controller-go-sdk/photon"
+	"gopkg.in/gcfg.v1"
+	"io"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	k8stypes "k8s.io/kubernetes/pkg/types"
+	"log"
+	"os/exec"
+	"strings"
+)
+
+const (
+	ProviderName = "photon"
+)
+
+// Global variable pointing to photon client, can be assigned to mock client in tests
+var Esxclient *photon.Client
+var logger *log.Logger = nil
+var overrideIP bool = false
+
+// Photon is an implementation of cloud provider Interface for Photon Controller.
+type PCCloud struct {
+	cfg *PCConfig
+	// InstanceID of the server where this PCCloud object is instantiated.
+	localInstanceID string
+	// local $HOSTNAME
+	localHostname string
+	// hostname from K8S, could be overridden
+	localK8sHostname string
+	// Photon Project ID
+	projID string
+	//	instance      *Instances
+	cloudprovider.Zone
+}
+
+type PCConfig struct {
+	Global struct {
+		CloudTarget       string `gcfg:"target"`
+		IgnoreCertificate bool   `gcfg:"ignoreCertificate"`
+		Tenant            string `gcfg:"tenant"`
+		Project           string `gcfg:"project"`
+		OverrideIP        bool   `gcfg:"overrideIP"`
+	}
+}
+
+// Disks is interface for manipulation with PhotonController Persistent Disks.
+type Disks interface {
+	// AttachDisk attaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	AttachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DetachDisk detaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	DetachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DiskIsAttached checks if a disk is attached to the given node.
+	DiskIsAttached(pdID string, nodeName k8stypes.NodeName) (bool, error)
+
+	// CreateDisk creates a new PD with given properties.
+	CreateDisk(volumeOptions *VolumeOptions) (pdID string, err error)
+
+	// DeleteDisk deletes PD.
+	DeleteDisk(pdID string) error
+}
+
+// VolumeOptions specifies capacity, tags, name and flavorID for a volume.
+type VolumeOptions struct {
+	CapacityGB int
+	Tags       map[string]string
+	Name       string
+	Flavor     string
+}
+
+func readConfig(config io.Reader) (PCConfig, error) {
+	if config == nil {
+		err := fmt.Errorf("no Photon Cloud Provider config file given")
+		return PCConfig{}, err
+	}
+
+	var cfg PCConfig
+	err := gcfg.ReadInto(&cfg, config)
+	return cfg, err
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(ProviderName, func(config io.Reader) (cloudprovider.Interface, error) {
+		cfg, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return newPCCloud(cfg)
+	})
+}
+
+func getVMIDbyNodename(project string, nodeName string) (string, error) {
```

Comment:
Function header comments would be nice. Does it matter if VM is powered off or on?